reconcile/blackbox_exporter_endpoint_monitoring.py

```diff
@@ -1,80 +1,53 @@
 import logging
 import sys
-from urllib.parse import urlparse
 from typing import Any, Optional
-import json
-from dataclasses import field
-
-from pydantic.dataclasses import dataclass
 
 from reconcile import queries
 from reconcile.utils.openshift_resource import (
-    OpenshiftResource, ResourceInventory)
-from reconcile.utils.defer import defer
+    OpenshiftResource)
 from reconcile.utils.semver_helper import make_semver
-import reconcile.openshift_base as ob
-
-QONTRACT_INTEGRATION = "blackbox-exporter-endpoint-monitoring"
-QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
-
-LOG = logging.getLogger(__name__)
-
-
-@dataclass(frozen=True, eq=True)
-class BlackboxMonitoringProvider:
-
-    module: str
-    # the namespace of a blackbox-exporter provider is mapped as dict
-    # since its only use with ob.fetch_current_state is as a dict
-    namespace: dict[str, Any] = field(compare=False, hash=False)
-    exporterUrl: str
-
 
-@dataclass(frozen=True, eq=True)
-class EndpointMonitoringProvider:
+from reconcile.closedbox_endpoint_monitoring_base import run_for_provider, EndpointMonitoringProvider, Endpoint, parse_prober_url
 
-    name: str
-    provider: str
-    description: str
-    timeout: Optional[str] = None
-    checkInterval: Optional[str] = None
-    blackboxExporter: Optional[BlackboxMonitoringProvider] = None
-    metricLabels: Optional[str] = None
-
-    @property
-    def metric_labels(self):
-        return json.loads(self.metricLabels) if self.metricLabels else {}
 
+QONTRACT_INTEGRATION = "blackbox-exporter-endpoint-monitoring"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
 
-@dataclass
-class Endpoint:
+PROVIDER = "blackbox-exporter"
 
-    name: str
-    description: str
-    url: str
+LOG = logging.getLogger(__name__)
 
-    @dataclass
-    class Monitoring:
 
-        provider: EndpointMonitoringProvider
+def run(dry_run: bool, thread_pool_size: int, internal: bool,
+        use_jump_host: bool) -> None:
+    # verify that only allowed blackbox-exporter modules are used
+    settings = queries.get_app_interface_settings()
+    allowed_modules = \
+        set(settings["endpointMonitoringBlackboxExporterModules"])
+    verification_errors = False
+    if allowed_modules:
+        for p in get_blackbox_providers():
+            if p.blackboxExporter and \
+                 p.blackboxExporter.module not in allowed_modules:
+                LOG.error(
+                    f"endpoint monitoring provider {p.name} uses "
+                    f"blackbox-exporter module {p.blackboxExporter.module} "
+                    f"which is not in the allow list {allowed_modules} of "
+                    "app-interface-settings"
+                )
+                verification_errors = True
+    if verification_errors:
+        sys.exit(1)
```

Comment:
I 100% agree but it can't be done here.

Here's the thing: `hack/run_integration.py` runs this integration, and there is an `except SystemExit` somewhere in there. Based on that `SystemExit` we increase the counters for successful or failed runs, which then alert us when integrations are truly hosed.

So, yes, I would **love** to get rid of all these `sys.exit`, but it needs to be done in a different place, and with good thought on how we're going to report if an integration is succeeding or not.

Reply:
i will not touch this for now. as @Piojo stated, this needs good thought and a general approach